File: src/main/java/net/lazars/garden/service/SchedulingService.java
package net.lazars.garden.service;

import jakarta.annotation.PostConstruct;
import java.time.Instant;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.TaskScheduler;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class SchedulingService {

  private final TaskScheduler taskScheduler;
  private final LightsService lightsService;
  private final SunriseSunsetService sunriseSunsetService;

  @PostConstruct
  void init() {
    scheduleTask(sunriseSunsetService.getSunsetData());
  }

  @Scheduled(cron = "${garden.lights.turn-off}", zone = "${garden.location.zone}")
  void turnOffLights() {
    lightsService.turnOff();
  }

  void onTaskRun() {
    log.info("Task start");
    lightsService.turnOn();

    final Instant upcomingSunset = sunriseSunsetService.loadUpcomingSunsetData();
    scheduleTask(upcomingSunset);
  }

  private void scheduleTask(final Instant instant) {
    taskScheduler.schedule(new TurnOnLightsTask(this), instant);
    log.info("Scheduled task for {}", instant);
  }

  private record TurnOnLightsTask(SchedulingService schedulingService) implements Runnable {

    @Override
    public void run() {
      schedulingService.onTaskRun();
    }
  }
}

File: src/main/kotlin/net/lazars/garden/service/SchedulingService.kt
package net.lazars.garden.service

import org.slf4j.LoggerFactory
import java.time.*
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class SchedulingService(
    private val lightsService: LightsService,
    private val sunriseSunsetService: SunriseSunsetService,
    private val lightsOffTime: LocalTime,
    private val zone: String
) {

    private val log = LoggerFactory.getLogger(javaClass)
    private val executor = Executors.newSingleThreadScheduledExecutor()

    init {
        scheduleTask(sunriseSunsetService.sunsetData) { _ -> executeLightsOnTask() }
        scheduleTask(getLightsOffTime()) { _ -> executeLightsOffTask() }
    }

    private fun scheduleTask(instant: Instant, task: (Any?) -> Unit) {
        val now = Instant.now()
        val duration = Duration.between(now, instant)

        executor.schedule({
            task.invoke(null)
        }, duration.toMillis(), TimeUnit.MILLISECONDS)
    }

    private fun executeLightsOnTask() {
        log.info("Lights on task start")
        lightsService.turnOn()

        scheduleTask(sunriseSunsetService.loadUpcomingSunsetData()) { _ -> executeLightsOnTask() }
    }

    private fun executeLightsOffTask() {
        log.info("Lights off task start")
        lightsService.turnOff()

        scheduleTask(getLightsOffTime()) { _ -> executeLightsOffTask() }
    }

    private fun getLightsOffTime(): Instant {
        var lightsOffTime = LocalDateTime.of(LocalDate.now(), lightsOffTime)
        if (!lightsOffTime.isAfter(LocalDateTime.now())) lightsOffTime = lightsOffTime.plusDays(1)

        val instant = lightsOffTime.atZone(ZoneId.of(zone)).toInstant()
        log.info("Lights off: $instant")

        return instant
    }
}
